functorch/csrc/DynamicLayer.cpp

```diff
@@ -544,21 +541,23 @@ void dynamicLayerFrontFallback(const c10::OperatorHandle& op, torch::jit::Stack*
   TORCH_INTERNAL_ASSERT(dynamicLayerStack.size() > 0);
   SaveLocalDispatchKeySet guard;
 
-  // if is a grad transform, and the operation is in-place, and the mutated
-  // argument is not currently wrapped in a TensorWrapper, then we need to
-  // error out otherwise the result is silently incorrect
-  checkForInvalidMutationOnCaptures(op, stack, dynamicLayerStack);
-
   // Unwrap dead GradWrappers, materialize live ones
   auto maybeTransformGradWrappers = [](const Tensor& tensor) {
     auto result = unwrapIfDead(tensor);
     return materializeGradWrappers(result, getDynamicLayerStack());
   };
-  auto num_args = op.schema().arguments().size();
-  foreachTensorInplace(*stack, stack->size() - num_args, stack->size(), maybeTransformGradWrappers);
 
+  auto num_args = op.schema().arguments().size();
   auto& layer = dynamicLayerStack.back();
 
+  // if is a grad transform, and the operation is in-place, and the mutated
+  // argument is not currently wrapped in a TensorWrapper, then we need to
+  // error out otherwise the result is silently incorrect
+  if (dynamicLayerStack.back().key() == DispatchKey::Autograd) {
+    checkForInvalidMutationOnCaptures(op, stack, dynamicLayerStack);
+    foreachTensorInplace(*stack, stack->size() - num_args, stack->size(), maybeTransformGradWrappers);
+  }
```

Comment:
The fix I'm applying is to "only bother running logic related to autograd if the autograd transform is active".

The bug was that if you run `grad(functionalize(f))`, we would end up with a `FunctionalTensorWrapper` that has *both* the `Functionalize` key and the `FuncTorchGradWrapper` key on it, because `FunctionalTensorWrapper` copies the keyset from its inner tensor.

You might ask why the functional wrapper copies all of the keys from the inner tensor into its keyset. I originally tried to fix this by having `FunctionalTensorWrapper` not copy all dispatch keys from its inner tensor, which sounds pretty reasonable. This is actually pretty fraught though, for a few reasons:

(1) In the LazyTensor/XLA case, we still want most functionality keys to be copied onto the wrapper tensor, because functionalities like autograd / autocast will actually run directly on the wrapper
(2) Maybe we could just copy all of the keys onto the wrapper except for the functorch keys. I ran into an issue though, with the `PythonTLSSnapshot` key (which is between the functorch layer keys). Basically, if the inner tensor is a python tensor, then we definitely want the snapshot key to *also* be on the wrapper.

Those two issues seem fixable... but "only run autograd logic if an autograd transform is active" seems simple and reasonable.

Reply:
Yeah, I agree